tests/src/test/java/io/streamnative/pulsar/handlers/kop/storage/ProducerStateManagerTest.java

```diff
@@ -0,0 +1,582 @@
+import java.util.Collections;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.InvalidTxnStateException;
+import org.apache.kafka.common.errors.OutOfOrderSequenceException;
+import org.apache.kafka.common.internals.Topic;
+import org.apache.kafka.common.record.ControlRecordType;
+import org.apache.kafka.common.record.EndTransactionMarker;
+import org.apache.kafka.common.record.RecordBatch;
+import org.mockito.Mockito;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+import org.testng.collections.Lists;
+
+/**
+ * Producer state manager test.
+ */
+@Slf4j
+public class ProducerStateManagerTest extends KopProtocolHandlerTestBase {
+
+    protected final long defaultTestTimeout = 20000;
+    private final TopicPartition partition = new TopicPartition("test", 0);
+    private final Long producerId = 1L;
+    private final MockTime time = new MockTime();
+    private ProducerStateManager stateManager;
+
+    @BeforeClass
+    @Override
+    protected void setup() throws Exception {
+        this.conf.setKafkaTransactionCoordinatorEnabled(false);
+        super.internalSetup();
+
+        admin.topics().createPartitionedTopic("public/default/sys-topic-producer-state", 1);
+        log.info("success internal setup");
+    }
+
+    @BeforeMethod
+    protected void setUp() {
+        stateManager = new ProducerStateManager(partition.toString());
+    }
+
+    @AfterMethod
+    protected void tearDown() {
+        // no-op
+    }
+
+    @AfterClass
+    @Override
+    protected void cleanup() throws Exception {
+        super.internalCleanup();
+    }
+
+    @Test(timeOut = defaultTestTimeout)
+    public void testBasicIdMapping() {
+        short epoch = 0;
+
+        // First entry for id 0 added
+        append(stateManager, producerId, epoch, 0, 0L, 0L, false);
+
+        // Second entry for id 0 added
+        append(stateManager, producerId, epoch, 1, 0L, 1L, false);
+
+        // Duplicates are checked separately and should result in OutOfOrderSequence if appended
+        try {
+            append(stateManager, producerId, epoch, 1, 0L, 1L, false);
+            fail("Duplicates are checked separately and should result in OutOfOrderSequence if appended.");
+        } catch (OutOfOrderSequenceException e) {
+            log.info("Expected behavior.");
+        }
+
+        // Invalid sequence number (greater than next expected sequence number)
+        try {
+            append(stateManager, producerId, epoch, 5, 0L, 2L, false);
+            fail("Invalid sequence number (greater than next expected sequence number).");
+        } catch (OutOfOrderSequenceException e) {
+            log.info("Expected behavior.");
```

Comment:
Please remove all "Expected behavior." logs because they don't help at all.

And I think it's better to use `assertThrows` for this case, like

```java
        assertThrows(OutOfOrderSequenceException.class, () ->
                append(stateManager, producerId, epoch, 5, 0L, 2L, false));
```